libminifi/include/core/ReferenceParser.h

```diff
@@ -0,0 +1,135 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "minifi-cpp/Exception.h"
+#include "utils/file/AssetManager.h"
+
+namespace org::apache::nifi::minifi::core {
+
+using IdResolver = std::function<std::optional<std::string>(std::string /* category */, std::string /* value */)>;
+
+class MalformedReferenceException : public Exception {
+ public:
+  explicit MalformedReferenceException(const std::string &message) : Exception(ExceptionType::FLOW_EXCEPTION, message) {}
+
+  explicit MalformedReferenceException(const char *message) : Exception(ExceptionType::FLOW_EXCEPTION, message) {}
+};
+
+class UnknownCategoryException : public Exception {
+ public:
+  explicit UnknownCategoryException(const std::string &message) : Exception(ExceptionType::FLOW_EXCEPTION, message) {}
+
+  explicit UnknownCategoryException(const char *message) : Exception(ExceptionType::FLOW_EXCEPTION, message) {}
+};
+
+std::string resolveIdentifier(const std::string &str, std::vector<IdResolver> resolvers) {
+  std::string result;
+  enum class ParseState {
+    OutsideToken,
+    InAtMark,
+    InToken,
+    InId
+  } state{ParseState::OutsideToken};
+  std::string category;
+  std::string id;
+  for (size_t offset = 0; offset < str.length(); ++offset) {
+    if (state == ParseState::OutsideToken) {
+      if (str[offset] == '@') {
+        state = ParseState::InAtMark;
+        continue;
+      }
+      result += str[offset];
+    } else if (state == ParseState::InAtMark) {
+      if (str[offset] == '@') {
+        result += '@';
+        state = ParseState::OutsideToken;
+        continue;
+      }
+      if (str[offset] == '{') {
+        state = ParseState::InToken;
+        category.clear();
+        id.clear();
+        continue;
+      }
+      throw MalformedReferenceException(fmt::format("Malformed reference at {} in '{}'", offset, str));
+    } else if (state == ParseState::InToken) {
+      if (str[offset] == ':') {
+        state = ParseState::InId;
+        continue;
+      }
+      category += str[offset];
+    } else if (state == ParseState::InId) {
+      if (str[offset] == '}') {
+        std::optional <std::string> replacement;
+        for (auto &resolver: resolvers) {
+          replacement = resolver(category, id);
+          if (replacement) {
+            break;
+          }
+        }
+        if (!replacement) {
+          throw UnknownCategoryException(fmt::format("Could not resolve '{}' in category '{}'", id, category));
+        }
+        result += replacement.value();
+        state = ParseState::OutsideToken;
+        continue;
+      }
+      id += str[offset];
+    }
+  }
+  if (state != ParseState::OutsideToken) {
+    throw MalformedReferenceException("Unterminated id");
```

Comment:
This message is over-specific, since we can also get here if the input contains "@{missing_category}", for example.  Something like
```suggestion
    throw MalformedReferenceException(fmt::format("Found invalid asset reference expression in '{}'", str));
```
would be better.